test/conformance/testing/include/uur/known_failure.h

```diff
@@ -0,0 +1,117 @@
+        : DeviceMatcher(2, UR_PLATFORM_BACKEND_LEVEL_ZERO,
+                        {il.begin(), il.end()}) {}
+};
+
+struct CUDA : DeviceMatcher {
+    CUDA(std::initializer_list<std::string> il)
+        : DeviceMatcher(1, UR_PLATFORM_BACKEND_CUDA, {il.begin(), il.end()}) {}
+};
+
+struct HIP : DeviceMatcher {
+    HIP(std::initializer_list<std::string> il)
+        : DeviceMatcher(1, UR_PLATFORM_BACKEND_HIP, {il.begin(), il.end()}) {}
+};
+
+struct NativeCPU : DeviceMatcher {
+    NativeCPU(std::initializer_list<std::string> il)
+        : DeviceMatcher(1, UR_PLATFORM_BACKEND_NATIVE_CPU,
+                        {il.begin(), il.end()}) {}
+};
+
+inline bool isKnownFailureOn(ur_adapter_handle_t adapter,
+                             ur_platform_handle_t platform,
+                             ur_device_handle_t device,
+                             std::vector<DeviceMatcher> deviceMatchers) {
+    for (const auto &deviceMatcher : deviceMatchers) {
+        uint32_t adapterVersion = 0;
+        urAdapterGetInfo(adapter, UR_ADAPTER_INFO_VERSION,
+                         sizeof(adapterVersion), &adapterVersion, nullptr);
+        ur_platform_backend_t backend = UR_PLATFORM_BACKEND_UNKNOWN;
+        uur::GetPlatformInfo<ur_platform_backend_t>(
+            platform, UR_PLATFORM_INFO_BACKEND, backend);
+        if (deviceMatcher.adapterVersion != adapterVersion &&
+            deviceMatcher.backend != backend) {
+            continue;
+        }
+        if (deviceMatcher.deviceNames.empty()) {
+            return true;
+        }
+        std::string deviceName;
+        uur::GetDeviceInfo<std::string>(device, UR_DEVICE_INFO_NAME,
+                                        deviceName);
+        for (const auto &unsupportedDeviceName : deviceMatcher.deviceNames) {
+            if (deviceName.find(unsupportedDeviceName) != std::string::npos) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+} // namespace uur
+
+#define UUR_KNOWN_FAILURE_ON(...)                                              \
+    if (uur::isKnownFailureOn(adapter, platform, device, {__VA_ARGS__})) {     \
+        using namespace std::literals;                                         \
+        std::string platformName;                                              \
+        uur::GetPlatformInfo<std::string>(platform, UR_PLATFORM_INFO_NAME,     \
+                                          platformName);                       \
+        std::string deviceName;                                                \
+        uur::GetDeviceInfo<std::string>(device, UR_DEVICE_INFO_NAME,           \
+                                        deviceName);                           \
+        const char *alsoRunKnownFailures =                                     \
+            std::getenv("UR_CTS_ALSO_RUN_KNOWN_FAILURES");                     \
+        if (alsoRunKnownFailures && (alsoRunKnownFailures == "1"sv ||          \
```

Comment:
Could getting and checking the environment variable be done in a separate function? Just to avoid having a huge macro in each failing test.

Reply:
I've done this in [another branch](https://github.com/oneapi-src/unified-runtime/commit/3d1a452dc704f96b394bce7a41147bf84400177a#diff-51b8d25935e4f911b926c9aa2a160ac6cc039206078953b1763c46d72b37a005R181) which is based on top of the work to support multiple adapters/paltforms/devices at once. This work will likely get rolled into those changes.